### ISSUE TITLE
Add mergeKey for PrometheusRule CRD

### ISSUE THREAD

**Issue by akamac** [prometheus-operator/prometheus-operator#5195]
**What is missing?**
mergeKey definition for rules in PrometheusRule CRD
**Why do we need it?**
To be able to override the specific rule with strategicMerge patch.

For now, only JSON patch can be used for this, but using rule's index in path (no selectors are allowed) is not descriptive.

**Comment by simonpasquier** [comment]
You mean mergeKey for groups?

https://github.com/prometheus-operator/prometheus-operator/blob/4e026ff8f9ccac7ab12a7118a94f316ae5e8f183/pkg/apis/monitoring/v1/types.go#L1820-L1823

mergeKey for individual rules isn't really possible with the current schema because of the difference between alerting and recording rules.

**Comment by akamac** [comment]
> You mean mergeKey for groups?
> 
> https://github.com/prometheus-operator/prometheus-operator/blob/4e026ff8f9ccac7ab12a7118a94f316ae5e8f183/pkg/apis/monitoring/v1/types.go#L1820-L1823
> 
> mergeKey for individual rules isn't really possible with the current schema because of the difference between alerting and recording rules.

Yeah, I've just realised that while patching the existing schema.
Merging only rule groups is not that interesting.

I guess we should add `name` field for both alerting and recording rules and make it mandatory, thus allowing to merge rules on `name` key.

**Comment by simonpasquier** [comment]
> I guess we should add name field for both alerting and recording rules and make it mandatory, thus allowing to merge rules on name key.

We can't do that without bumping the API version unfortunately.
